teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,124 @@
+---
+title: TaskRun Breakpoint on Failure of Step 
+authors:
+  - "@waveywaves"
+creation-date: 2021-01-15
+last-updated: 2021-03-21
+status: proposed
+---
+
+# TEP-0042: TaskRun Breakpoint on Failure of Step 
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+Debugging TaskRuns can be tiresome. Re-running Tasks to figure out which part of a particular step is to blame would tax productivity.
+By enabling breakpoint on failure for a TaskRun we should be able to halt a TaskRun at the failing step and get access to the
+step environment to analyze cause of the failure. Unlike legacy systems, this allows the user to debug TaskRuns while the TaskRun 
+is still running, hence improving developer productivity and Pipeline debuggability.
+
+## Motivation
+
+Lack of live debugging for Pipelines.
+
+### Goals
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  generateName: quarante-deux
+spec:
+  breakpoint:
+    onFailure: true
```

Comment:
i like that we're leaving this open to adding more configuration 👍 

i've got a couple nitpicks about the format here but i think they'll be worth it in the long run!

1. I'm thinking maybe we can call the section `debug` instead of `breakpoint`? I'm thinking about how I'm used to interacting with this kind of thing, e.g. this menu in goland:

![image](https://user-images.githubusercontent.com/432502/110811245-85e25b00-8254-11eb-84fb-e2ae1413f79a.png)

It has "run" vs "debug", and even an action like "attach to process" has the same icon as "debug", i.e. "debugging" is kind of the main activity

2. Instead of a 'true/false' boolean let's try to use a string type (e.g. [see this k8s guidance about avoiding bools](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types)

If we combine those two together, we can get something like:

```yaml
debug:
   break: [onFailure] # i'm jumping to a conclusion that we might want this to be a list, e.g. more than one condition under which we want to break - but maybe i'm getting ahead of myself!!
```

If you're aware of other conditions under which we might want to break in the future maybe we can prepare for them in this design im thinking like:

* on failure
* at some specific point (e.g. at the start? when a particular step runs?)
* ???? when some condition is met? probably don't want to go that far....

Reply:
@bobcatfish thank you for putting your time in reviewing this proposal :) 

> i like that we're leaving this open to adding more configuration ￼

Awesome :)

> I'm thinking maybe we can call the section debug instead of breakpoint? I'm thinking about how I'm used to interacting with this kind of thing, e.g. this menu in goland:

I agree, it does make more sense to call it debug. Provides enough information to the user to not get confused by the terminology and is closer to the usual developer routine.

> Instead of a 'true/false' boolean let's try to use a string type (e.g. see this k8s guidance about avoiding bools
If we combine those two together, we can get something like:
>```yaml
>debug:
>   break: [onFailure] # i'm jumping to a conclusion that we might want this to be a list, e.g. more than one condition under >which we want to break - but maybe i'm getting ahead of myself!!
>```

After using debug I am expecting the spec to look something like below

```yaml
debug:
  breakpoint: 
    onFailure: true
    beforeStep: []
    afterStep: []
```

`breakpoint` instead of `break` sounds less ambiguous to me and aligns with the debug terminology as well. WDYT ?

The above also shows an idea of what it could look like if the user could provide more points where they can halt the TaskRun i.e. before and after certain steps. I didn't add it to this doc or make it part of this discussion initially to avoid a debate between referencing the steps nominally or ordinally :stuck_out_tongue: 

I wanted to keep the `onFailure: true` as it seems like you should be able to switch the flip on and off and would be an easy way to make sure that we would be able to debug the moment the breakpoint fails. It should also make sense to raise an event on the failure of a step (through a conditon probably ???). 

There might come a time, if rerunning steps does become a reality (which is highly likely), the user would like to debug before the failure takes place by enabling, something like `beforeFailure` which would rerun the step and halt it before execution.

```yaml
debug:
  breakpoint: 
    onFailure: 
      before: true
      after: true
    beforeStep: []
    afterStep: []
```

Maybe I am getting ahead of myself but this does seem like a possibility at hand and I do swing back to your idea of having a list instead of a boolean.

```yaml
debug:
  breakpoint: ["onFailure", "beforeFailure"]
```

If we had to add before and after steps in the equation, it could look something like

```yaml
debug:
  breakpoint: ["before-push-to-registry", "after-3"]
```

where the user provides the location before or after the step which they want to debug followed by the name or the index number of the step itself.

P.S.
Considering we already teased the topic above, it would be nice to provide the ability to reference a step either by their name or index number (should the starting index be 0 or 1 (I think it should be 1 because it would be the "first step")).

>If you're aware of other conditions under which we might want to break in the future maybe we can prepare for them in this design im thinking like:
> - on failure
> - at some specific point (e.g. at the start? when a particular step runs?)
> - ???? when some condition is met? probably don't want to go that far....

Hope the above provides enough detail and designs we can think over before finalizing.

Reply:
> breakpoint instead of break sounds less ambiguous to me and aligns with the debug terminology as well. WDYT ?

Nice improvement! looks good to me :D

> There might come a time, if rerunning steps does become a reality (which is highly likely), the user would like to debug before the failure takes place by enabling, something like beforeFailure which would rerun the step and halt it before execution.

Oh that's an interesting scenario - so in that case, Tekton would have to know where the failure happened last time and could preemptively break before it hits that same spot again? (We don't need to get into this now since I'm sure  this is a future enhancement XD but im curious if you've seen a precedent for this with other debugging tools?)

```
debug:
  breakpoint: ["onFailure", "beforeFailure"]
```

i like this approach - even if we only support `breakpoint: ["onFailure"]` for now i think it leaves us open to a lot of other options in the future

> it would be nice to provide the ability to reference a step either by their name or index number. Naturally the next question would be..should the starting index be 0 or 1 for that I think it should be 1 because it IS the "first" step that runs.

interesting point! hahaha oh no, i hope we go with 0 😆  but also i think we can avoid that debate until later!!

>Hope the above provides enough detail and designs we can think over before finalizing.

That's great!!! Thanks for looking at this so thoroughly - I think a lot of this would be great to include in the proposal itself, e.g. as "notes" or even as "alternatives"